rpc: add optional blockhash to waitfornewblock, unhide wait methods in help

Sjors opened the pull request:
The `waitfornewblock` is inherently racy as the tip may have changed since the last RPC call, and can even change during initial processing of this call.

Add an optional `blockhash` argument so the caller can specify their current tip. Return immediately if our tip is different.

I've made it fail if `LookupBlockIndex` fails. This should never happen if the user got the block hash from our RPC in the first place.

Finally, the `waitfor{block,newblock,blockheight}` RPC methods are no longer hidden in `help`:
- the changes in #30409 ensured these methods _could_ work in the GUI
- #31785 removed the guards that prevented GUI users from using them
- this PR makes `waitfornewblock` reliable
 
So there's no more reason to hide them.

Sjors commented:
Now that #30409 landed this is a single commit, and ready for review.

Sjors commented on src/rpc/blockchain.cpp:
```diff
@@ -283,7 +284,22 @@ static RPCHelpMan waitfornewblock()
+        }
+    }
+
     if (IsRPCRunning()) {
```
So this effects GUI users that use the console, rather than bitcoin-cli?

ryanofsky commented on src/rpc/blockchain.cpp:
```diff
@@ -283,7 +284,22 @@ static RPCHelpMan waitfornewblock()
     NodeContext& node = EnsureAnyNodeContext(request.context);
     Mining& miner = EnsureMining(node);
 
+   // Use tip as reference block, unless a block hash was provided.
     auto block{CHECK_NONFATAL(miner.getTip()).value()};
+    if (!request.params[1].isNull()) {
+        uint256 hash(ParseHashV(request.params[1], "blockhash"));
+        {
+            ChainstateManager& chainman = EnsureAnyChainman(request.context);
+            LOCK(cs_main);
+            const CBlockIndex* block_index{chainman.m_blockman.LookupBlockIndex(hash)};
+            if (!block_index) {
+                throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, "Block not found");
+            }
+            block.hash = block_index->GetBlockHash();
+            block.height = block_index->nHeight;
+        }
+    }
+
     if (IsRPCRunning()) {
         block = timeout ? miner.waitTipChanged(block.hash, std::chrono::milliseconds(timeout)) : miner.waitTipChanged(block.hash);
     }
```
I think it would be good to simplify all of this to just:

```c++
    // Use tip as reference block, unless a block hash was provided.
    uint256 current_tip{request.params[1].isNull()
        ? miner.getTip().value_or(BlockRef{}).hash
        : ParseHashV(request.params[1], "blockhash")};

    BlockRef block{timeout > 0
        ? miner.waitTipChanged(current_tip, std::chrono::milliseconds(timeout))
        : miner.waitTipChanged(current_tip)};
```

In addition to being simpler, this would fix a number of problems and inconveniences of the current implementation:

- If an unrecognized block hash was passed, it would make the RPC return the current block tip instead of throwing an exception.
- If no block hash was passed and tip was null, it would make the RPC wait for a block to be connected instead of throwing an exception.
- It would fix the bug marco described https://github.com/bitcoin/bitcoin/pull/30635#discussion_r1778228488 where this function does not wait in the GUI console.

ryanofsky reviewed (approved):
Code review ACK 13b867fb9b37fef3f14732904172e4ecba6fc973. I think implementation could be a ilttle simpler and more friendly though (see suggestion below)

luke-jr commented on src/rpc/blockchain.cpp:
```diff
@@ -283,7 +284,22 @@ static RPCHelpMan waitfornewblock()
     auto block{CHECK_NONFATAL(miner.getTip()).value()};
+    if (!request.params[1].isNull()) {
+        uint256 hash(ParseHashV(request.params[1], "blockhash"));
+        {
+            ChainstateManager& chainman = EnsureAnyChainman(request.context);
+            LOCK(cs_main);
+            const CBlockIndex* block_index{chainman.m_blockman.LookupBlockIndex(hash)};
+            if (!block_index) {
+                throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, "Block not found");
+            }
+            block.hash = block_index->GetBlockHash();
+            block.height = block_index->nHeight;
+        }
```
Why do all this? Can't we just compare `hash` to `block` and return?

Sjors replied:
@luke-jr it should be simpler now

Sjors commented:
@ryanofsky & @luke-jr thanks for reviewing, will address feedback soon(tm).

Sjors commented:
I took @ryanofsky's suggestion: https://github.com/bitcoin/bitcoin/pull/30635#discussion_r1792082432

ryanofsky commented on src/rpc/blockchain.cpp:
```diff
@@ -283,10 +285,14 @@ static RPCHelpMan waitfornewblock()
-    if (IsRPCRunning()) {
-        block = timeout ? miner.waitTipChanged(block.hash, std::chrono::milliseconds(timeout)) : miner.waitTipChanged(block.hash);
-    }
+    // Use tip as reference block, unless a block hash was provided.
+    uint256 current_tip{request.params[1].isNull()
+        ? miner.getTip().value_or(BlockRef{}).hash
+        : ParseHashV(request.params[1], "current_tip")};
```
In commit "rpc: add optional blockhash to waitfornewblock" (c305ad5b06545497752b2f12d8d9f72e8a6c2941)

Might be useful if comment explained the different cases here. Maybe: // If caller provided a current_tip value, pass it to waitTipChanged. If the caller did not provide a current tip hash, call getTip() to get one and wait for the tip to be different from this value. (This mode is less reliable because if the tip changed between waitfornewblock calls, it will need to change a second time before this call returns.) If getTip() returns null, which can happen during startup, pass 0 to waitTipChanged() so it will wait for any tip hash to be set.

ryanofsky commented on src/rpc/blockchain.cpp:
```diff
@@ -283,10 +285,14 @@ static RPCHelpMan waitfornewblock()
     Mining& miner = EnsureMining(node);
 
-    auto block{CHECK_NONFATAL(miner.getTip()).value()};
-    if (IsRPCRunning()) {
```
In commit "rpc: add optional blockhash to waitfornewblock" (c305ad5b06545497752b2f12d8d9f72e8a6c2941)

Note: dropping IsRPCRunning call is a minor change in behavior, but probably a good one, as it should let the waitfornewblock function in the GUI console even if the RPC server is not enabled. In any case it is safe to drop this because waitTipChanged already will return if the node is shutting down.

Sjors replied:
Good point. Just tested that that's indeed the case. I'll mention that in the commit message.

cc @hebasto

ryanofsky reviewed (approved):
Code review ACK c305ad5b06545497752b2f12d8d9f72e8a6c2941. This is a nice change that should make the RPC (and test) more reliable, and the implementation seems to be pretty simple now.

Sjors committed:
rpc: add optional blockhash to waitfornewblock

Sjors commented:
Rebased, added comment and expanded commit message to note the GUI improvement.

Sjors commented:
I added a commit to make `waitfornewblock` visible in `help`. It was introduced in 2016 by #8680 as a quick fix to deal with CI failures.

At this point is seem mature enough, no longer has weird quirks like requiring GUI users to enable the RPC server, it's probably actually useful and at least [one library uses it anyway](https://github.com/rust-bitcoin/rust-bitcoincore-rpc/blob/06a6c156b4d469c60c1abe8afdc002e7e11e5d3b/client/src/client.rs#L1100). So it seems fine to list it.

sedited reviewed (approved):
ACK 1f8808d411fc0ed2b91f185e08ad8c1ee8f8432f

Tested in the GUI too.

ryanofsky reviewed (approved):
Code review ACK 1f8808d411fc0ed2b91f185e08ad8c1ee8f8432f

Since last review a comment was expanded, and a new commit makes the `waitfornewblock` RPC not hidden anymore

Sjors commented:
Rebased after #31600.

ryanofsky reviewed (approved):
Code review ACK 3d88cad169aa3d2d8731c3888749054be6688f1b. No changes since last review other than rebase due to conflict in a test.

Sjors commented:
#31785 increasingly overlaps with the scope of this PR, so I'm going to rebase this on it.

Sjors committed:
rpc: unhide waitfor{block,newblock,blockheight}

They are now reliable. An earlier commit dropped their
IsRPCRunning() guards so they also work in the GUI.

Sjors commented:
This is now based on #31785, and we now unhide all three wait methods.

ryanofsky reviewed (approved):
Code review ACK 798bc8d400b354d626678dc7b54c3cb0f464aa7a. Changes since last review were rebasing on top of #31785 to simplify this PR, updating comments and release notes, and also unhiding waitforblock and waitforblockheight methods

ryanofsky reviewed (approved):
Code review ACK e47b20f2f310f05832c879401660860cc40a6a09 with no changes since last review other than rebase

ryanofsky commented on src/rpc/blockchain.cpp:
```diff
@@ -287,10 +288,26 @@ static RPCHelpMan waitfornewblock()
+     * less reliable because if the tip changed between waitfornewblock calls,
+     * it will need to change a second time before this call returns.
+     *
+     * Abort if RPC came out of warmup too early.
```
In commit "rpc: add optional blockhash to waitfornewblock" (0625f26b09084fcbd4daa48b3456f54eff6b22de)

Do we think "Abort if RPC came out of warmup too early." comment is still accurate? I've having trouble understanding what it could mean and I don't think there are cases where this aborted even before this change. Would probably suggest dropping this comment or trying to make it clearer.

Also could replace /** with a normal comment above since /** normally indicates a doxygen comment.

Sjors replied:
I replaced the `/**`.

I dropped the warmup comment, but will keep the `CHECK_NONFATAL`. If the RPC starts listening before `ActiveChain().Tip()` is set, then `getTip()` returns nothing. I'm not sure if that can ever happen. `SetRPCWarmupFinished()` is called as step 13 in `init.cpp` with a somewhat ominous comment above it.

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/30635#discussion_r2052789582

Thanks, I think I get it now. "Abort if RPC came out of warmup too early" meant "There is a bug if RPC came out of warmup too early, so abort". But it should be pretty obvious that if a check statement fails, there is a bug, so the comment seems redundant and your change dropping the comment and keeping the check makes sense.

ryanofsky commented on doc/release-30635.md:
```diff
@@ -1,4 +1,5 @@
 ------------
 
-- The waitfornewblock RPC takes an optional `current_tip` argument. (#30635)
+- The waitfornewblock takes an optional `current_tip` argument. It is also no longer hidden. (#30635)
```
In commit "rpc: unhide waitfor{block,newblock,blockheight}" (067acb5907b5b135cff702da13a1f64a02d64110)

Maybe s/takes/now takes/ to be clear this is a change

ryanofsky reviewed (approved):
Code review ACK 067acb5907b5b135cff702da13a1f64a02d64110. No changes since previous review other than rebasing

sedited reviewed (approved):
Re-ACK c6e2c31c55123cc97b4400bcbf3c37a39b067a22

ryanofsky commented on doc/release-30635.md:
```diff
@@ -1,4 +1,5 @@
 ------------
 
-- The waitfornewblock RPC takes an optional `current_tip` argument. (#30635)
+- The waitfornewblock now takes an optional `current_tip` argument. It is also no longer hidden. (#30635)
```
In commit "rpc: unhide waitfor{block,newblock,blockheight}" (c6e2c31c55123cc97b4400bcbf3c37a39b067a22)

Accidentally dropped the word RPC?

Sjors replied:
It seems so. Will fix release note if I have to retouch.

ryanofsky reviewed (approved):
Code review ACK c6e2c31c55123cc97b4400bcbf3c37a39b067a22. Just rebased and tweaked documentation since last review.

glozow reviewed (commented):
utACK c6e2c31c55123cc97b4400bcbf3c37a39b067a22
